src/Tasks.UnitTests/AssemblyDependency/ResolveAssemblyReferenceTestFixture.cs

```diff
@@ -300,6 +305,7 @@ internal void StartIOMonitoring()
             // If tables are present then the corresponding IO function will do some monitoring.
             uniqueFileExists = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
             uniqueGetAssemblyName = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+            uniqueGetDirectoryFiles = new ConcurrentDictionary<string, int>(StringComparer.OrdinalIgnoreCase);
```

Comment:
I believe that current implementation is not thread safe. Default xunit execution runs tests in derived classes concurrently and since this (uniqueFileExists, uniqueGetAssemblyName  and uniqueGetDirectoryFiles) are static, they can be damaged by concurrent write. 
In theory it could be that only one class actually write to those but to be sure we would have to carefully review all 20ish derived classes.
I did wanted to make situation with thread safety worse, so I have decided to plat safe and use ConcurrentDictionary

Reply:
This should guarantee that tests in the same assembly never run in parallel, making statics and other shared state legal:

https://github.com/dotnet/msbuild/blob/f98579d50f8b08c4efe13ca07327e3e5222b50f8/src/Shared/UnitTests/TestAssemblyInfo.cs#L11

I would personally lean towards not using `ConcurrentDictionary` to avoid making a false impression that tests are thread-safe. On the other hand I accept your argument of not making it worse, so fine either way.

Reply:
+1

Reply:
I was looking for it, but I have missed it, maybe because it is included in `src\Directory.Build.targets` as conditional item.
Changed to `Dictionary`.